# Homework #2

How to do it? Uploading your .ipynb file into the corresponding homework in the LMS system. Any delay will not be accepted. 

##  Question 1: (Total 6 points)

Below is a mapping from city abbreviations to their Chinese city names 

In [1]:
city="""TPE＝台北
TXG＝台中
KHH＝高雄
HUN＝花蓮
HKG＝香港
MFM＝澳門
BJS＝北京
SHA＝上海
CAN＝廣州
CTU＝成都
CKG＝重慶
CGO＝鄭州
CSX＝長沙
DLC＝大連
FOC＝福州
HAK＝海口
HFE＝合肥
HGH＝杭州
HRB＝哈爾濱
KWE＝貴陽
KMG＝昆明
KWL＝桂林
NGB＝寧波
NKG＝南京
NNG＝南寧
SHE＝瀋陽
SYX＝三亞
TAO＝青島
TNA＝濟南
TSN＝天津
WNZ＝溫州
WUH＝武漢
WUX＝無錫
SIA＝西安
XMN＝廈門
TYO＝東京
OSA＝大阪
NGO＝名古屋
OKA＝琉球
FUK＝福岡
SEL＝首爾
PUS＝釜山
CJU＝濟州島
BKK＝曼谷
HKT＝普吉島
CNX＝清邁
SIN＝新加坡
KUL＝吉隆坡
KCH＝古晉
PEN＝檳城
JKT＝雅加達
DPS＝巴里島
SUB＝泗水
SGN＝胡志明市
HAN＝河內
DAD＝峴港
MNL＝馬尼拉
CEB＝宿霧
LAO＝佬沃
ADL＝阿德雷得
BNE＝布里斯班
CBR＝坎培拉
CNS＝凱恩斯
MEL＝墨爾本
PER＝伯斯
SYD＝雪梨
AKL＝奧克蘭
CHC＝基督城
DEL＝德里
BOM＝孟買
LON＝倫敦
BHX＝伯明罕
EDI＝愛丁堡
MAN＝曼徹斯特
DUB＝都柏林
PAR*＝巴黎
NCE＝尼斯
BER＝柏林
DUS＝杜塞道夫
FRA＝法蘭克福
HAM＝漢堡
HAJ＝漢諾威
MUC＝慕尼黑
NUE＝紐倫堡
STR＝斯圖加特
WAW＝華沙
AMS＝阿姆斯特丹
ZRH＝蘇黎世
GVA＝日內瓦
VIE＝維也納
SZG＝薩爾斯堡
MAD＝馬德里
BCN＝巴塞隆納
LIS＝里斯本
ROM＝羅馬
MIL＝米蘭
VCE＝威尼斯
ATH＝雅典
YOW＝渥太華
YTO＝多倫多
YVR＝溫哥華
NYC＝紐約
WAS＝華盛頓
LAX＝洛杉磯
SFO＝舊金山
SEA＝西雅圖
ATL＝亞特蘭大
AUS＝奧斯汀
BOS＝波士頓
CHI＝芝加哥
DEN＝丹佛
DTT＝底特律
HNL＝夏威夷檀香山
HOU＝休斯頓
MIA＝邁阿密
PHL＝費城
PHX＝鳳凰城
MEX＝墨西哥
PTY＝巴拿馬"""

### (a) Make a mapping table (2 points)  

In [2]:
# Method 1 by lhlee： data structure only 
cr = city.replace("\n","＝")
CodeName = cr.split("＝")
ECDict = dict(zip(CodeName[::2],CodeName[1::2]))
ECDict.items()

dict_items([('TPE', '台北'), ('TXG', '台中'), ('KHH', '高雄'), ('HUN', '花蓮'), ('HKG', '香港'), ('MFM', '澳門'), ('BJS', '北京'), ('SHA', '上海'), ('CAN', '廣州'), ('CTU', '成都'), ('CKG', '重慶'), ('CGO', '鄭州'), ('CSX', '長沙'), ('DLC', '大連'), ('FOC', '福州'), ('HAK', '海口'), ('HFE', '合肥'), ('HGH', '杭州'), ('HRB', '哈爾濱'), ('KWE', '貴陽'), ('KMG', '昆明'), ('KWL', '桂林'), ('NGB', '寧波'), ('NKG', '南京'), ('NNG', '南寧'), ('SHE', '瀋陽'), ('SYX', '三亞'), ('TAO', '青島'), ('TNA', '濟南'), ('TSN', '天津'), ('WNZ', '溫州'), ('WUH', '武漢'), ('WUX', '無錫'), ('SIA', '西安'), ('XMN', '廈門'), ('TYO', '東京'), ('OSA', '大阪'), ('NGO', '名古屋'), ('OKA', '琉球'), ('FUK', '福岡'), ('SEL', '首爾'), ('PUS', '釜山'), ('CJU', '濟州島'), ('BKK', '曼谷'), ('HKT', '普吉島'), ('CNX', '清邁'), ('SIN', '新加坡'), ('KUL', '吉隆坡'), ('KCH', '古晉'), ('PEN', '檳城'), ('JKT', '雅加達'), ('DPS', '巴里島'), ('SUB', '泗水'), ('SGN', '胡志明市'), ('HAN', '河內'), ('DAD', '峴港'), ('MNL', '馬尼拉'), ('CEB', '宿霧'), ('LAO', '佬沃'), ('ADL', '阿德雷得'), ('BNE', '布里斯班'), ('CBR', '坎培拉'), ('CNS', '凱恩斯'), ('MEL', '墨爾本'), ('PER', '伯

In [3]:
# Method 2 by lhlee： with program structure 
ECDict = {} # empty dictionary
cr = city.replace("\n","＝")
CodeName = cr.split("＝")
for i in range(len(CodeName[::2])): # CodeName[::2]: city code list
    c = CodeName[::2][i]
    ECDict[c]= CodeName[1::2][i] # CodeName[1::2]: Chinese name list
#ECDict.items()

In [4]:
# Method 3 by lhlee： dictionary comprehension (generator)
cr = city.replace("\n","＝")
CodeName = cr.split("＝")
ECDict = { CodeName[::2][i]:CodeName[1::2][i] for i in range(len(CodeName[::2]))}
# ECDict.items()

In [5]:
#By 昌浩
#split the city
city_list=city.split('\n')
#split the city abbreviation  and chinese name
citysplit=[]
for i in city_list:
    citysplit.append(i.split('＝'))
#make the mapping table
import pandas as pd
citydf=pd.DataFrame(citysplit,columns=['abbr','Chinese'])
citydf

abbr Chinese
0    TPE      台北
1    TXG      台中
2    KHH      高雄
3    HUN      花蓮
4    HKG      香港
..   ...     ...
114  MIA     邁阿密
115  PHL      費城
116  PHX     鳳凰城
117  MEX     墨西哥
118  PTY     巴拿馬

[119 rows x 2 columns]

In [6]:
#By 柏翰
a = city.split()
abr = []
chn = []
for i in a:
    abr.append(i.split('＝')[0])
    chn.append(i.split('＝')[1])
import pandas as pd
df = pd.DataFrame(list(zip(abr, chn)), columns =['abr', 'chn'])
print(df)

     abr  chn
0    TPE   台北
1    TXG   台中
2    KHH   高雄
3    HUN   花蓮
4    HKG   香港
..   ...  ...
114  MIA  邁阿密
115  PHL   費城
116  PHX  鳳凰城
117  MEX  墨西哥
118  PTY  巴拿馬

[119 rows x 2 columns]


### (b) Given an inputted city abbreviation and then check whether it exists or not? If yes, showing its Chinese city name. (4 points)

In [7]:
# Method 1 by lhlee: if .. else ..
code = input("Please input a city abbreviation: ")
if code in ECDict:
    print(code, ", its Chinese city name is ", ECDict[code])
else:
    print(code, "NOT in the mapping table")
    

Please input a city abbreviation: TPE
TPE , its Chinese city name is  台北


In [8]:
# Method 2 by lhlee: while ...
while True:
    code = input("Please input a city abbreviation (type q to quit)：")
    if (code =="q"):
        break
    if code in ECDict:
        print(code, ", its Chinese city name is ", ECDict[code])
    else:
        print(code, "NOT in the mapping table")

Please input a city abbreviation (type q to quit)：TPE
TPE , its Chinese city name is  台北
Please input a city abbreviation (type q to quit)：XXX
XXX NOT in the mapping table
Please input a city abbreviation (type q to quit)：q


In [9]:
#By 昌浩
def show(df,city):
    abbr=df.abbr.tolist()
    citylist=df.Chinese.tolist()
    if city in abbr:
        print('City abbreviation:',city)
        print('exists')
        print('Chinese city name:',citylist[abbr.index(city)])
        return True
    else:        
        print('City abbreviation:',city)
        print('NOT exists')
        return False

while True:
    abbr_=input("Given an inputted city abbreviation :")
    abbr_=abbr_.upper() #將輸入轉成大寫
    check=show(citydf,abbr_)
    if check==False:
        break

Given an inputted city abbreviation :TPE
City abbreviation: TPE
exists
Chinese city name: 台北
Given an inputted city abbreviation :XXX
City abbreviation: XXX
NOT exists


## Question 2: (4 Points)

Below is a mapping from a case-insensitive character to its corresponding number. 

### What is the sum of the corresponding numbers in the poem《The Zen of Python》written by Tim Peters? 

In [10]:
poem ="""The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!"""

In [11]:
#Method 1: Counter 

from collections import Counter

character = [chr(ord("a")+i) for i in range(0, 26)] # List comprehension
score = dict(zip(character, range(1,27))) # Dict{'a':1, 'b':2, .. 'z':26}

output=0
letters = [x for x in list(poem.lower()) if x in character] 
letter_dict = dict(Counter(letters)) # Dict{'letter': count}
for k, v in letter_dict.items():
    output += score[k]*v
output

8045

In [12]:
# Method 2: List Comprehension

character = [chr(ord("a")+i) for i in range(0, 26)] # List comprehension
score = dict(zip(character, range(1,27))) # Dict{'a':1, 'b':2, .. 'z':26}
output = [score[x] for x in list(poem.lower()) if x in character]
sum(output)

8045

In [13]:
#By 昌浩
import string
charl=[i for i in string.ascii_lowercase] #小寫字母
charu=[i for i in string.ascii_uppercase] #大寫字母
num =[i for i in range(1,27)] #數字1-26
char=list(zip(charl,charu,num))
print('map:',char)
count=0
for i in poem:
    if i in charl :
        count+=num[charl.index(i)]
    elif i in charu:
        count+=num[charu.index(i)]
print('The sum of the corresponding numbers in the poem《The Zen of Python》written by Tim Peters:',count)

map: [('a', 'A', 1), ('b', 'B', 2), ('c', 'C', 3), ('d', 'D', 4), ('e', 'E', 5), ('f', 'F', 6), ('g', 'G', 7), ('h', 'H', 8), ('i', 'I', 9), ('j', 'J', 10), ('k', 'K', 11), ('l', 'L', 12), ('m', 'M', 13), ('n', 'N', 14), ('o', 'O', 15), ('p', 'P', 16), ('q', 'Q', 17), ('r', 'R', 18), ('s', 'S', 19), ('t', 'T', 20), ('u', 'U', 21), ('v', 'V', 22), ('w', 'W', 23), ('x', 'X', 24), ('y', 'Y', 25), ('z', 'Z', 26)]
The sum of the corresponding numbers in the poem《The Zen of Python》written by Tim Peters: 8045


In [14]:
#By 柏翰
want = []
lower = ['a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z']
upper = ['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z']
sum = 0
for i in range(26):
    want.append(int(i+1))

for i in range(26):
    sum += list(poem).count(lower[i])*want[i]
    sum += list(poem).count(upper[i])*want[i]
print(sum)

8045


# 評分標準
-------------------------

## 配分

### 　
* Answer1 
##### (a) mapping table 為119項的 dict 或 dataframe (2分)
##### (b) 1.檢查是否存在 (2分) 　2.是否對應正確中文城市名稱　(2分)

### 　
* Answer2 = 8045 (4分)
#####  直接print答案的數字 0分



### 沒有學號或是命名不符合規則扣1分